### PR TITLE
Make message export not rely on message keys being in the MessageStoreCache.

### PR DESCRIPTION
**Pull request by smn**
This assumption goes away with the work going on in praekelt/vumi#675. Currently the message store cache is being used as a convenient way of retrieving message keys and timestamps, an alternative to that needs to be found.

### CODE REVIEW

**Comment by smn** [comment]
@hodgestar @justinvdm ready for review

File: go/conversation/tasks.py [hunk 1]
```diff
@@ -8,29 +10,124 @@
...
+def email_export(user_profile, conversation, io):
+    zipio = StringIO()
+    zf = ZipFile(zipio, "a", ZIP_DEFLATED)
+    zf.writestr("messages-export.csv", io.getvalue())
+    zf.close()
+
+    email = EmailMessage(
+        'Conversation message export: %s' % (conversation.name,),
+        'Please find the messages of the conversation %s attached.\n' % (
+            conversation.name),
```
Comment (justinvdm) [inline go/conversation/tasks.py:75]: If `email_export` is a reusable method for sending out emails, should we maybe have the email subject and contents be params instead? The current subject and contents look specific to message exports.

Comment (smn) [inline go/conversation/tasks.py:75]: Happy for suggestions. I did consider that but I decided to keep this usable within the context of these tasks, trying to make it more generic only gave me a very thin layer on top of the already existing email tools that Django gives us with little added benefit. Perhaps I should rename it or prefix it with an underscore?

Comment (justinvdm) [inline go/conversation/tasks.py:75]: Ah, looks like all of the tasks are actually message related, happy with it staying the way it is.

File: go/conversation/tests.py [hunk 2]
```diff
@@ -839,3 +842,62 @@ def bad_add_entity(*a, **kw):
...
+            email.recipients(), [self.user_helper.get_django_user().email])
+        self.assertTrue(conv.name in email.subject)
+        self.assertTrue(conv.name in email.body)
+        [(file_name, zipcontent, mime_type)] = email.attachments
+        self.assertEqual(file_name, 'messages-export.zip')
+        zipfile = ZipFile(StringIO(zipcontent), 'r')
+        content = zipfile.open('messages-export.csv', 'r').read()
+        # 1 header, 5 sent, 5 received, 1 trailing newline == 12
+        self.assertEqual(12, len(content.split('\n')))
+        self.assertEqual(mime_type, 'application/zip')
```
Comment (justinvdm) [inline go/conversation/tests.py]: I'm wondering if it would be helpful to check that we are exporting the messages we are expecting. Checking based on lines in the file is probably enough, but checking against the actual messages feels like a stronger assertion. Could we not do something similar to the `export_conversation_messages_sorted` test, but sort the result rows when doing the assertion?

**Comment by justinvdm** [comment]
Other than the really minor comment about the `export_conversation_messages_unsorted` test, looks good. I do think someone else should take a look though, I'm probably not familiar enough with the message store to know if things are being done right/wrong.

**Comment by smn** [comment]
Ready for re-review.

**Comment by justinvdm** [comment]
I'm :+1: on this landing, but I'm not too confident on my knowledge of the message store. Your call if you think someone else needs to review it.

**Comment by smn** [comment]
@justinvdm thanks! @jerith @hodgestar for when you have time.

File: go/conversation/tasks.py [hunk 3]
```diff
@@ -8,29 +10,124 @@
...
+    scrubbing of messages depending on `include_sensitive` and `scrubber`.
+
+    :param Conversation conv:
+        The conversation.
+    :param str direction:
+        The direction, either ``'inbound'`` or ``'outbound'``.
+    :param int size:
+        How big the chunks should be. Default 20.
+    :param bool include_sensitive:
+        If ``true`` then all messages marked as `sensitive` are skipped.
```
Comment (jerith) [inline go/conversation/tasks.py]: The param name and description say opposite things. I think the description is wrong, unless `collect_messages()` (which we pass the value to) also uses a bad name.

File: go/conversation/tasks.py [hunk 4]
```diff
@@ -8,29 +10,124 @@
...
+    :param callable scrubber:
+        If provided, this is called for every message allowing it to be
+        modified on the fly.
+    """
+    direction = ('inbound' if direction == 'inbound' else 'outbound')
+    keys_func = getattr(conversation, '%s_keys' % (direction,))
+    proxy = getattr(conversation.mdb, '%s_messages' % (direction,))
```
Comment (jerith) [inline go/conversation/tasks.py]: The direction should be either `'inbound'` or `'outbound'` here and we should throw a `ValueError` for anything that isn't. I'd also prefer to explicitly name the attributes we're getting (in an `if` block or something) instead of constructing their names and using `getattr()`.

Comment (jerith) [inline go/conversation/tasks.py]: Actually, since we're calling `keys_func` exactly once just below this, it might make sense to just call the right thing instead of sticking it in a local variable first.

File: go/conversation/tasks.py (hunk 3 above)
Comment (smn) [inline go/conversation/tasks.py]: Thanks, fixed.

File: go/conversation/tasks.py [hunk 5]
```diff
@@ -8,29 +10,124 @@
...
+
+    keys = keys_func()
+    for chunk in grouper(keys, size):
+        # grouper() pads with `None` if less than `size` available,
+        # we unpad here.
+        chunk = filter(None, chunk)
```
Comment (jerith) [inline go/conversation/tasks.py]: Perhaps it would make sense to skip `grouper()` and just use `izip_longest()` here directly without worrying about the padding we're adding and then removing.

File: go/conversation/tasks.py (hunk 4 above)
Comment (smn) [inline go/conversation/tasks.py]: You're right, fixed.

File: go/conversation/tasks.py [hunk 6]
```diff
@@ -8,29 +10,130 @@
...
+    """
+    if direction == 'inbound':
+        keys = conversation.inbound_keys()
+        proxy = conversation.mdb.inbound_messages
+    elif direction == 'outbound':
+        keys = conversation.outbound_keys()
+        proxy = conversation.mdb.outbound_messages
+    else:
+        raise ValueError('Invalid value received for `direction`. '
+                         'Only `inbound` and `outbound` are allowed.')
```
Comment (jerith) [inline go/conversation/tasks.py]: Perhaps put the value we got in the exception? That might be handy for debugging. Also, the `direction` param should now either be mandatory or have a valid default value.

File: go/conversation/tasks.py (hunk 5 above)
Comment (smn) [inline go/conversation/tasks.py]: `izip_longest()` defaults to using `None` as the `fillvalue` and doesn't provide an easy means of grouping not seeing how using it directly here would make this cleaner?

File: go/conversation/tasks.py [hunk 7]
```diff
@@ -8,29 +10,130 @@
...
+
+    for chunk in grouper(keys, size):
+        # grouper() pads with `None` if less than `size` available,
+        # we unpad here.
+        chunk = filter(None, chunk)
+        messages = conversation.collect_messages(
+            chunk, proxy, include_sensitive=include_sensitive,
+            scrubber=scrubber)
+        yield messages
```
Comment (jerith) [inline go/conversation/tasks.py]: Instead of doing all this here, it might be better to factor the message processing out of `collect_messages()` and just use `load_all_bunches()` on the proxy object directly. We should actually get rid of `collect_messages()` completely, because it always loads all messages in a conversation into memory and that's going to hurt us no matter where it's used.

File: go/conversation/tasks.py [hunk 8]
```diff
@@ -8,29 +10,131 @@
...
+
+    email_export(user_profile, conversation, io)
+
+
+@task(ignore_result=True)
+def export_conversation_messages(account_key, conversation_key):
+    warnings.warn('export_conversation_messages() is deprecated. '
+                  'Please use export_conversation_messages_sorted() instead',
+                  category=DeprecationWarning)
+    return export_conversation_messages_sorted(account_key, conversation_key)
```
Comment (jerith) [inline go/conversation/tasks.py]: Do we need to deprecate this if we find all the places that call it and change them?

File: go/conversation/tasks.py (hunk 7 above)
Comment (smn) [inline go/conversation/tasks.py]: `collect_messages()` is useful for the UI but needs to be used with caution, if you give it all the keys it will load them all into memory. The point is not to do that. It is also used by `export_conversation_messages_sorted()` which _is_ currently loading all messages in memory for sorting (which is why I'm doing this work). I'm happy to make that go away in a separate PR.

File: go/conversation/tasks.py (hunk 8 above)
Comment (smn) [inline go/conversation/tasks.py]: My thinking is that I'm not sure who else is currently using this task and where, out side of us.

File: go/conversation/tests.py [hunk 9]
```diff
@@ -839,3 +842,65 @@ def bad_add_entity(*a, **kw):
 
         self.assertEqual(self.error_log, [':('])
         self.assertEqual(response.context['dashboard_config'], None)
+
+
+class TestConversationTasks(GoDjangoTestCase):
+    def setUp(self):
+        self.vumi_helper = self.add_helper(
+            DjangoVumiApiHelper(), setup_vumi_api=False)
+        self.vumi_helper.setup_vumi_api()
```
Comment (jerith) [inline go/conversation/tests.py]: Why pass `setup_vumi_api=False` and then immediately call `setup_vumi_api()`?

File: go/vumitools/conversation/utils.py [hunk 10]
```diff
@@ -258,9 +259,17 @@ def collect_messages(self, keys, proxy, include_sensitive, scrubber):
...
         returnValue(collection)
 
-    @Manager.calls_manager
     def received_messages(self, start=0, limit=100, include_sensitive=False,
                           scrubber=None):
+        warnings.warn('received_messages() is deprecated. Please use '
+                      'received_messages_in_cache() instead.',
+                      category=DeprecationWarning)
+        return self.received_messages_in_cache(start, limit, include_sensitive,
+                                               scrubber)
```
Comment (jerith) [inline go/vumitools/conversation/utils.py:283]: I'd rather find and fix all the places that call this now than deprecate it. There isn't any third-party code relying on it.

File: go/conversation/tests.py (hunk 9 above)
Comment (smn) [inline go/conversation/tests.py]: copy paste error, thanks, fixing.

File: go/conversation/tasks.py (hunk 7 above)
Comment (jerith) [inline go/conversation/tasks.py]: `collect_messages()` isn't particularly useful for the UI because showing a hundred thousand messages isn't feasible in any case. I'm happy to have a ticket for removing it instead of doing it in this branch, but I think it's still worth factoring out the message processing part of it and using that with `load_all_bunches()` to simplify the logic here.

File: go/vumitools/conversation/utils.py (hunk 10 above)
Comment (smn) [inline go/vumitools/conversation/utils.py:283]: Done & in a single, hopefully won't meddle with reviewability too much.

File: go/conversation/tasks.py (hunk 8 above)
Comment (jerith) [inline go/conversation/tasks.py]: We make breaking changes in Vumi Go all the time, so I don't think it's worth doing the deprecation thing here unless we also decide to do it for all the other internal things. I think the cost of that is much too high when there's no evidence that anyone other than us is using the code.

Comment (jerith) [inline go/conversation/tasks.py]: I'm happy to keep the deprecation stuff here in the short term if celery needs it, though. I don't know how it handles code changing underneath scheduled tasks.

File: go/conversation/tasks.py (hunk 7 above)
Comment (smn) [inline go/conversation/tasks.py]: It's what drives the loading of messages to show your messages in the paginated message list, we do this by only handing a subset of all keys to `collect_messages()` (only the keys for that page). It is probably worth factoring out but there's still going to be shared code between this & the UI as both need to allow for the scrubbing & skipping of sensitive messages.

File: go/conversation/tasks.py (hunk 8 above)
Comment (smn) [inline go/conversation/tasks.py]: Ok, removed.

File: go/conversation/tasks.py (hunk 7 above)
Comment (jerith) [inline go/conversation/tasks.py]: I meant factor it out into a thing on `ConversationWrapper` that takes a list of messages (or just one message, if that makes more sense) and processes it.

Comment (smn) [inline go/conversation/tasks.py]: Cool, can address that in a separate ticket.

File: go/conversation/tasks.py (hunk 5 above)
Comment (jerith) [inline go/conversation/tasks.py]: Ah, good point.

**Comment by jerith** [comment]
I'm happy with this branch once my commits have been reviewed and Travis has done its thing successfully.
